Autotools/vpath support

kleetus opened the pull request:
closes #10
## Discussion

LevelDB currently has a hand-rolled Makefile and it seems to work well, especially for build systems that do not have a shell (Windows). What the build system for LevelDB does not have is support for Autotools, specifically leveraging autconf, autoreconf, and the macros that come along with that. 

LevelDB is, fundamentally, a project that should be a dependency to another project with its own build system. Many of these projects are also open-source and use Autotools. As such, those projects can easily integrate this project if, and only if, this project has at least a configure.ac file. Otherwise those projects need to hack in running "make -c" in their own Makefile.am's.
## This code adds
- a basic configure.ac and autogen.sh script
- a basic configure script
- a Makefile input file that mirrors the original Makefile, so that non-shell systems can still continue to use the Makefile

Commit:
Added a basic structure for autoreconf to work with.

Commit:
Enabled out of tree builds

- added macro in main configure.ac for running configure script for leveldb
- added autogen script to launch autoreconf in leveldb
- added the most basic configure.ac for leveldb
- added references to srcdir instead of hardcoded path

laanwj commented:
Concept ACK

da2ce7 commented:
I don't think that the generated configure script should be included in the pull request.

Maybe update the .gitignore?

laanwj commented on .gitignore:
```diff
@@ -11,3 +11,7 @@ Release
+autom4te.cache/
+config.log
+config.status
+!configure
```
Agree that a generated script like configure shouldn't be in the repo. A typo, I think.

sipa replied:
The pull request lists it as a feature, so I assume it's intentional.

kleetus commented:
The addition of the configure script is intentional. The rationale is consistency with secp256k1 library which includes a configure shell script. Without the configure script, parent projects would need to run autoconf or autoreconf on leveldb.

laanwj commented:
> Without the configure script, parent projects would need to run autoconf or autoreconf on leveldb.

The problem with committing build-generated scripts is that when developers do run `autoconf`, a repository file will have changes, which is easy to inadvertently commit, resulting in merge conflicts with other developers (as well as a lot of commit noise).

(as a convenience, ./configure script is part of the tarball created with `make dist`, but not the repository)

Also I suppose the only intended parent project here is bitcoin, which will handle calling the autoconf.

laanwj commented:
Tested ACK apart from the configure nit - see https://github.com/bitcoin/bitcoin/issues/7466#issuecomment-182408931

sipa commented:
There is no ./configure script included in the secp256k1 repository, by the way: https://github.com/bitcoin/secp256k1 nor https://github.com/bitcoin/bitcoin/tree/master/src/secp256k1

kleetus commented:
> The problem with committing build-generated scripts is that when developers do run autoconf, a repository file will have changes, which is easy to inadvertently commit, resulting in merge conflicts with other developers (as well as a lot of commit noise).

Agreed! Let me fix this so that there is no configure script and that gitignore is properly aligned to ignore the output of autoreconf.

Commit:
Removed configure script (this will be generated by autoconf). Adjusted .gitignore to ignore this file.

Commit:
1. Removed target distclean and maintainer clean (unneeded currently). 
2. Added a slash in Makefile input so as to not modify the main Makefile when configure is run. The result is a double slash when referring to paths in the build, but this is harmless.

laanwj commented:
@theuni can you take a look here please?

theuni commented:
Hmm, I wonder if it would be simpler to just add our own simple Makefile.am here, rather than trying to pile on their stuff. I'll whip something up for comparison.

laanwj commented:
Well this is already done, and checked it works.

theuni commented:
Pushed an alternative to https://github.com/theuni/bitcoin/commit/3b714691956e623935929e5b174eceb8ab342bcd

The benefit is that it's directly integrated into our build without the need for another configure run. I'm not married to it, but I'm not a fan of adding more complexity to what should be a pretty straightforward build.

laanwj commented:
I agree, it's a worthwhile alternative - however, please read the discussion in https://github.com/bitcoin/bitcoin/issues/7466 on how we got to this solution.

I think the advantage of your solution is that it can be maintained entirely outside of leveldb's tree (given that the .include is moved). Effectively it absorbs leveldb into bitcoin's build system.
On the other hand this solution would be better if leveldb upstream could be convinced to switch. It's more of a modular approach.

Not wedded to either though, I just want out of tree builds to work.

laanwj commented:
@theuni I've tested your patch, but it fails on `test_bitcoin` and the bench.

Seems every executable now needs both `$(INT_LIBLEVELDB) $(INT_LIBMEMENV)` as well as `$(LIBLEVELDB) $(LIBMEMENV)` in its LDADD?

theuni commented:
@laanwj I've pushed a cleaned up version here: https://github.com/theuni/bitcoin/tree/leveldb-integration. Having done that cleanly now, I do prefer it over the wrapper approach. That's not to diminish @kleetus's work, of course.

If you're ok with the approach, I'll PR it here.

laanwj commented:
@theuni Looks good to me! You probably want to pr it at bitcoin/bitcoin though, not here.

theuni commented:
Right, I forgot the file was moved up. PRing over there.

laanwj commented:
Closing this in favor of the build system changes in Bitcoin Core.

It may still make sense to try to do this upstream.
